gem/refactorise.py

```diff
@@ -112,6 +116,175 @@ def product(*args):
+        """
+        Group and factorise monomials based on a list of atomics. Create new monomials for
+        each group and optimise them recursively.
+        :param optimal_atomics: list of tuples of optimal atomics and their sum indices
+        :return: new MonomialSum object with atomics factorised.
+        """
+        if not optimal_atomics:
+            return self
+        if len(self.ordering) < 2:
+            return self
+        new_monomial_sum = MonomialSum()
+        new_monomial_sum.argument_indices = self.argument_indices
+        # Group monomials according to each optimal atomic
+        factor_group = OrderedDict()
+        for key, (_sum_indices, _atomics) in iteritems(self.ordering):
+            for (sum_indices_set, sum_indices), oa in optimal_atomics:
+                if key[0] == sum_indices_set and oa in _atomics:
+                    # Add monomial key to the list of corresponding optimal atomic
+                    factor_group.setdefault(((sum_indices_set, sum_indices), oa), []).append(key)
+                    break
+            else:
+                # Add monomials that do no have argument factors to new MonomialSum
+                new_monomial_sum.add(_sum_indices, _atomics, self.monomials[key])
+        # We should not drop monomials
+        assert sum([len(x) for x in itervalues(factor_group)]) + len(new_monomial_sum.ordering) == len(self.ordering)
+
+        for ((sum_indices_set, sum_indices), oa), keys in iteritems(factor_group):
+            if len(keys) == 1:
+                # Just one monomials with this atomic, add to new MonomialSum straightaway
+                _, _atomics = self.ordering[keys[0]]
+                _rest = self.monomials[keys[0]]
+                new_monomial_sum.add(sum_indices, _atomics, _rest)
+                continue
+            all_atomics = []  # collect all atomics from monomials
+            all_rest = []  # collect all rest from monomials
+            for key in keys:
+                _, _atomics = self.ordering[key]
+                _atomics = list(_atomics)
+                _atomics.remove(oa)  # remove common factor
+                all_atomics.append(_atomics)
+                all_rest.append(self.monomials[key])
+            # Create new MonomialSum for the factorised out terms
+            sub_monomial_sum = MonomialSum()
+            sub_monomial_sum.argument_indices = self.argument_indices
+            for _atomics, _rest in zip(all_atomics, all_rest):
+                sub_monomial_sum.add((), _atomics, _rest)
+            sub_monomial_sum = sub_monomial_sum.optimise()
+            assert len(sub_monomial_sum.ordering) != 0
+            if len(sub_monomial_sum.ordering) == 1:
+                # result is a product, add to new MonomialSum directly
+                (_, new_atomics), = list(itervalues(sub_monomial_sum.ordering))
+                new_atomics += (oa,)
+                new_rest, = list(itervalues(sub_monomial_sum.monomials))
```

Comment:
Why this obsession with casting perfectly good generators to lists before destructuring?